### PR TITLE
Disambiguate calli thunks based on full calling conventions

### LINKED ISSUE #153: Disambiguate CalliMarshallingMethodThunk based on modopts
Name mangling doesn't take modopts into account so we end up with two different signatures having the same simple name.

We should probably drop all modopt except for the one specifying calling convention.

Might need to add code to respect the calling convention too.

Test: baseservices/callconvs/TestCallingConventions

### PR DESCRIPTION
**Pull request by jkotas**
Fixes dotnet/runtimelab#153

### CODE REVIEW

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/TypeLoaderExceptionHelper.cs [hunk 1]
```diff
@@ -88,6 +88,14 @@ private static string GetFormatString(ExceptionStringID id)
                     return SR.InvalidProgram_Vararg;
                 case ExceptionStringID.InvalidProgramCallVirtFinalize:
                     return SR.InvalidProgram_CallVirtFinalize;
+                case ExceptionStringID.InvalidProgramNonStaticMethod:
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/TypeLoaderExceptionHelper.cs:91]: Unrelated bug found by TestCallingConventions test.

**Comment by jkotas** [comment]
cc @dotnet/ilc-contrib

File: src/coreclr/tools/Common/TypeSystem/Interop/UnmanagedCallingConventions.cs [hunk 2]
```diff
@@ -193,6 +194,53 @@ private static UnmanagedCallingConventions AccumulateCallingConventions(Unmanage
...
+            Debug.Assert(index == count);
+
+            return ret;
+
+            static EmbeddedSignatureData CreateCallConvEmbeddedSignatureData(TypeSystemContext context, string name)
+                => new()
+                {
+                    index = MethodSignature.IndexOfCustomModifiersOnReturnType,
+                    kind = EmbeddedSignatureDataKind.OptionalCustomModifier,
+                    type = context.SystemModule.GetKnownType("System.Runtime.CompilerServices", name)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Interop/UnmanagedCallingConventions.cs:240]: Nit: In theory we could have a corelib that doesn't define these but still uses the calling convention (just encoding them oldstyle). It's probably fine to have it fail and ask the user to add it to corelib.

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

I've lifted the smoke test from the other pull request so that we also have a smoke test for this

File: src/coreclr/tools/Common/TypeSystem/Interop/UnmanagedCallingConventions.cs (hunk 2 above)
Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Interop/UnmanagedCallingConventions.cs:240]: Yep

**Comment by jkotas** [comment]
Linux arm failure is #73247

**Pull request merged**